lib/Backend/GlobOpt.cpp

```diff
@@ -19198,30 +19198,54 @@ GlobOpt::OptDstIsInvariant(IR::RegOpnd *dst)
 }
 
 void
-GlobOpt::OptHoistToLandingPadUpdateValueType(
-    BasicBlock* landingPad,
+GlobOpt::OptHoistUpdateValueType(
+    Loop* loop,
     IR::Instr* instr,
-    IR::Opnd* opnd,
+    IR::Opnd* srcOpnd,
     Value* opndVal)
 {
-    if (instr->m_opcode == Js::OpCode::FromVar)
+    if (opndVal == nullptr || instr->m_opcode == Js::OpCode::FromVar)
     {
         return;
     }
 
-    Sym* opndSym = opnd->GetSym();;
+    Sym* opndSym = srcOpnd->GetSym();;
 
     if (opndSym)
     {
-        if (opndVal == nullptr)
-        {
-            opndVal = FindValue(opndSym);
-        }
-
+        BasicBlock* landingPad = loop->landingPad;
         Value* opndValueInLandingPad = FindValue(landingPad->globOptData.symToValueMap, opndSym);
         Assert(opndVal->GetValueNumber() == opndValueInLandingPad->GetValueNumber());
 
-        opnd->SetValueType(opndValueInLandingPad->GetValueInfo()->Type());
+        ValueType opndValueTypeInLandingPad = opndValueInLandingPad->GetValueInfo()->Type();
+
+        if (srcOpnd->GetValueType() != opndValueTypeInLandingPad)
+        {
+            if (instr->m_opcode == Js::OpCode::SetConcatStrMultiItemBE)
```

Comment:
If there is bailout to ensure string type is also hoisted, the value type of src1 for `SetConcatStrMultiItemBE` should be `string`, then a new `Conv_PrimStr` will not be prepended here.

---
In reply to: [104359344](https://github.com/Microsoft/ChakraCore/pull/2632#discussion_r104359344) [](ancestors = 104359344)